# 01. What is Numpy

NumPy is the fundamental package for scientific computing with Python. 
It is a package that provides high-performance vector, matrix and higher-dimensional data structures for Python. 
It is implemented in C and Fortran so when calculations are **vectorized**, performance is very good.

So, in a nutshell:

* a powerful Python extension for N-dimensional array
* a tool for integrating C/C++ and Fortran code
* designed for scientific computation: linear algebra and Signal Analysis

If you are a MATLAB&reg; user we recommend to read [Numpy for MATLAB Users](http://www.scipy.org/NumPy_for_Matlab_Users) and [Benefit of Open Source Python versus commercial packages](http://www.scipy.org/NumPyProConPage). 

I'm a supporter of the **Open Science Movement**, thus I humbly suggest you to take a look at the [Science Code Manifesto](http://sciencecodemanifesto.org/)

# Getting Started with Numpy Arrays

NumPy's main object is the **homogeneous** ***multidimensional array***. It is a table of elements (usually numbers), all of the same type. 

In Numpy dimensions are called **axes**. 

The number of axes is called **rank**. 

The most important attributes of an ndarray object are:

* **ndarray.ndim**     - the number of axes (dimensions) of the array. 
* **ndarray.shape**    - the dimensions of the array. For a matrix with n rows and m columns, shape will be (n,m). 
* **ndarray.size**     - the total number of elements of the array. 
* **ndarray.dtype**    - numpy.int32, numpy.int16, and numpy.float64 are some examples. 
* **ndarray.itemsize** - the size in bytes of elements of the array. For example, elements of type float64 has itemsize 8 (=64/8) 

To use `numpy` need to import the module it using of example:

In [3]:
import numpy as np  # naming import convention

### Terminology Assumption

In the `numpy` package the terminology used for vectors, matrices and higher-dimensional data sets is *array*. 

### Reference Documentation

* On the web: [http://docs.scipy.org](http://docs.scipy.org)/

* Interactive help:

In [4]:
np.array?

Docstring:
array(object, dtype=None, *, copy=True, order='K', subok=False, ndmin=0,
      like=None)

Create an array.

Parameters
----------
object : array_like
    An array, any object exposing the array interface, an object whose
    __array__ method returns an array, or any (nested) sequence.
dtype : data-type, optional
    The desired data-type for the array.  If not given, then the type will
    be determined as the minimum type required to hold the objects in the
    sequence.
copy : bool, optional
    If true (default), then the object is copied.  Otherwise, a copy will
    only be made if __array__ returns a copy, if obj is a nested sequence,
    or if a copy is needed to satisfy any of the other requirements
    (`dtype`, `order`, etc.).
order : {'K', 'A', 'C', 'F'}, optional
    Specify the memory layout of the array. If object is not an array, the
    newly created array will be in C order (row major) unless 'F' is
    specified, in which case it will be in Fortran order (c

If you're looking for something

## Numpy Array Object

`NumPy` has a multidimensional array object called ndarray. It consists of two parts as follows:
   
   * The actual data
   * Some metadata describing the data
    
    
The majority of array operations leave the raw data untouched. The only aspect that changes is the metadata.

<img src="images/ndarray_with_details.png" />

## Creating `numpy` arrays

There are a number of ways to initialize new numpy arrays, for example from

* a Python list or tuples
* using functions that are dedicated to generating numpy arrays, such as `arange`, `linspace`, etc.

### From lists

For example, to create new vector and matrix arrays from Python lists we can use the `numpy.array` function.

In [8]:
v = np.array([1, 2, 3, 4])
v

array([1, 2, 3, 4])

In [7]:
M = np.array([[1, 2], [3, 4]])
M

array([[1, 2],
       [3, 4]])

The `v` and `M` objects are both of the type `ndarray` that the `numpy` module provides.

In [9]:
print('Type of v: ', type(v))
print('Type of M: ', type(M))

Type of v:  <class 'numpy.ndarray'>
Type of M:  <class 'numpy.ndarray'>


The difference between the `v` and `M` arrays is only their shapes. 

To do so, we could use the `numpy.shape` function:

In [10]:
print('Shape of v: ', np.shape(v))
print('Shape of M: ', np.shape(M))

Shape of v:  (4,)
Shape of M:  (2, 2)


Alternatively, we can get information about the shape of an array by using the `ndarray.shape` **property** :

In [11]:
v.shape, M.shape

((4,), (2, 2))

Equivalently, we can get information about the **size** of the two `ndarrays`, namely the *total number of elements* in the array.

In [12]:
print('Size of v:', v.size)
print('Size of M:', M.size)

Size of v: 4
Size of M: 4


#### More properties of the `numpy array`

In [13]:
M.itemsize # bytes per element

4

In [14]:
M.nbytes # number of bytes

16

In [15]:
M.ndim # number of dimensions

2

## Using array-generating functions

For larger arrays it is inpractical to initialize the data manually, using explicit python lists. 

Instead we can use one of the many **functions** in `numpy` that generates arrays of different forms. 

Some of the more common are: 

* `np.arange`; 
* `np.linspace`; 
* `np.logspace`; 
* `np.mgrid`;
* `np.random.rand`;
* `np.diag`;
* `np.zeros`;
* `np.ones`;
* `np.empty`;
* `np.tile`.

### `np.arange`

In [16]:
x = np.arange(0, 10, 1) 
print(x)

[0 1 2 3 4 5 6 7 8 9]


In [17]:
# floating point step-wise range generatation
x = np.arange(-1, 1, 0.1)  
print(x)

[-1.00000000e+00 -9.00000000e-01 -8.00000000e-01 -7.00000000e-01
 -6.00000000e-01 -5.00000000e-01 -4.00000000e-01 -3.00000000e-01
 -2.00000000e-01 -1.00000000e-01 -2.22044605e-16  1.00000000e-01
  2.00000000e-01  3.00000000e-01  4.00000000e-01  5.00000000e-01
  6.00000000e-01  7.00000000e-01  8.00000000e-01  9.00000000e-01]


### `np.linspace` and `np.logspace`

In [18]:
# using linspace, both end points **ARE included**
np.linspace(0, 10, 25)

array([ 0.        ,  0.41666667,  0.83333333,  1.25      ,  1.66666667,
        2.08333333,  2.5       ,  2.91666667,  3.33333333,  3.75      ,
        4.16666667,  4.58333333,  5.        ,  5.41666667,  5.83333333,
        6.25      ,  6.66666667,  7.08333333,  7.5       ,  7.91666667,
        8.33333333,  8.75      ,  9.16666667,  9.58333333, 10.        ])

←←←←←←←←←←←←←←←←←←←←←←←←←←←←←←←←←←← stopped here (search for 'np.logspace' in notes)

In [19]:
np.logspace(0, np.e**2, 10, base=np.e)

array([1.00000000e+00, 2.27278564e+00, 5.16555456e+00, 1.17401982e+01,
       2.66829540e+01, 6.06446346e+01, 1.37832255e+02, 3.13263169e+02,
       7.11980032e+02, 1.61817799e+03])

### `np.mgrid`

In [ ]:
x, y = np.mgrid[0:5, 0:5]  # similar to meshgrid in MATLAB

In [ ]:
x

In [ ]:
y

### `np.random.rand` & `np.random.randn`

In [ ]:
import numpy as np
np.random.seed(4)

In [ ]:
# uniform random numbers in [0,1]
R = np.random.rand(5,5)

In [ ]:
R

In [ ]:
R

In [ ]:
# standard normal distributed random numbers
np.random.randn(5,5)

### `np.diag`

In [ ]:
# a diagonal matrix
np.diag([1,2,3])

In [ ]:
# diagonal with offset from the main diagonal
np.diag([1,2,3], k=-1)[::-1]

### `np.eye`

In [ ]:
# a diagonal matrix with ones on the main diagonal
np.eye(3, dtype='int')  # 3 is the 

### `np.zeros` and `np.ones`

In [ ]:
np.zeros((3,3))

In [ ]:
np.ones((3, 3))

### DIY

***Try by yourself*** the following commands:

    np.zeros((3,4))
    np.ones((3,4))
    np.empty((2,3))
    np.eye(5)
    np.diag(np.arange(5))
    np.tile(np.array([[6, 7], [8, 9]]), (2, 2))

## So, why is it useful then?

So far the `numpy.ndarray` looks awefully much like a Python **list** (or **nested list**). 

*Why not simply use Python lists for computations instead of creating a new array type?*

There are several reasons:

* Python lists are very general. 
    - They can contain any kind of object. 
    - They are dynamically typed. 
    - They do not support mathematical functions such as matrix and dot multiplications, etc. 
    - Implementing such functions for Python lists would not be very efficient because of the dynamic typing.
    
    
* Numpy arrays are **statically typed** and **homogeneous**. 
    - The type of the elements is determined when array is created.
    
    
* Numpy arrays are memory efficient.
    - Because of the static typing, fast implementation of mathematical functions such as multiplication and addition of `numpy` arrays can be implemented in a compiled language (C and Fortran is used).

In [ ]:
import numpy as np

In [ ]:
L = range(100000)

In [ ]:
%timeit [i**2 for i in L]

In [ ]:
a = np.arange(100000)

In [ ]:
%timeit a**2

In [ ]:
%timeit [element**2 for element in a]

---

## Exercises

### Simple arrays

* Create simple one and two dimensional arrays. First, redo the examples
from above. And then create your own.

* Use the functions `len`, `shape` and `ndim` on some of those arrays and
observe their output.

### Creating arrays using functions

* Experiment with `arange`, `linspace`, `ones`, `zeros`, `eye` and `diag`.

* Create different kinds of arrays with random numbers.

* Try setting the seed before creating an array with random values 
    - *hint*: use `np.random.seed`

* Look at the function `np.empty`. What does it do? When might this be
useful?

---

## Basic Data Types

    bool             | This stores boolean (True or False) as a bit

    inti             | This is a platform integer (normally either int32 or int64)
    int8             | This is an integer ranging from -128 to 127
    int16            | This is an integer ranging from -32768 to 32767
    int32            | This is an integer ranging from -2 ** 31 to 2 ** 31 -1
    int64            | This is an integer ranging from -2 ** 63 to 2 ** 63 -1
    
    uint8            | This is an unsigned integer ranging from 0 to 255
    uint16           | This is an unsigned integer ranging from 0 to 65535
    uint32           | This is an unsigned integer ranging from 0 to 2 ** 32 - 1
    uint64           | This is an unsigned integer ranging from 0 to 2 ** 64 - 1

    float16          | This is a half precision float with sign bit, 5 bits exponent, and 10 bits mantissa
    float32          | This is a single precision float with sign bit, 8 bits exponent, and 23 bits mantissa
    float64 or float | This is a double precision float with sign bit, 11 bits exponent, and 52 bits mantissa
    complex64        | This is a complex number represented by two 32-bit floats (real and imaginary components)
    complex128       | This is a complex number represented by two 64-bit floats (real and imaginary components)
    (or complex)


## Numerical Types and Representation

The **numerical dtype** of an array should be selected very carefully, as it directly affects the numerical representation of elements, that is: 

   * the number of **bytes used; 
   * the *numerical range*

So, then: **What happens if I try to represent a number that is Out of range?**

Let's have a go with **integers**, i.e., `int8` and `uint8`

In [ ]:
x = np.zeros(4, 'int8')  # Integer ranging from -128 to 127
x

In [ ]:
x[0] = 127
x

In [ ]:
x[0] = 128
x

In [ ]:
x[1] = 129
x

In [ ]:
x[2] = 257  # i.e. (128 x 2) + 1
x

In [ ]:
ux = np.zeros(4, 'uint8')  # Integer ranging from 0 to 255
ux

In [ ]:
ux[0] = 255
ux[1] = 256
ux[2] = 257
ux[3] = 513  # (256 x 2) + 1
ux

---

# Exercises - Basic Numpy

## Ex 1. Create an array containing integers from $2$ to $2^6$

#### Hint: use Python `range` function

## 1.1 Print `ndarray` attributes and properties
(e.g. `type`, `dtype`, `shape...`) using previous on

## Ex 2. Create a 3x3 Matrix array and fill it with integer numbers

## Ex3: Create a Matrix of any size and fill it with random numbers

### HInt: take a look at `numpy.random.rand`

---

# Indexing

### Setting up the data

In [ ]:
np.random.seed(42)  # Setting the random seed

In [ ]:
# a vector: the argument to the array function is a Python list
v = np.random.rand(10)
v

In [ ]:
# a matrix: the argument to the array function is a nested Python list
M = np.random.rand(10, 2)
M

We can index elements in an array using the square bracket and indices:

In [ ]:
# v is a vector, and has only one dimension, taking one index
v[0]

In [ ]:
# M is a matrix, or a 2 dimensional array, taking two indices 
M[1,1]

If we omit an index of a multidimensional array it returns the whole row (or, in general, a N-1 dimensional array) 

In [ ]:
M[1] 

The same thing can be achieved with using `:` instead of an index: 

In [ ]:
M[1,:] # row 1

In [ ]:
M[:,1] # column 1

We can assign new values to elements in an array using indexing:

In [ ]:
M[0,0] = 1

In [ ]:
M

In [ ]:
# also works for rows and columns
M[1,:] = 0
M[:,1] = -1

In [ ]:
M

## Index slicing

Index slicing is the technical name for the syntax `M[lower:upper:step]` to extract part of an array:

In [ ]:
a = np.array([1,2,3,4,5])
a

In [ ]:
a[1:3]

Array slices are **mutable**: if they are assigned a new value the original array from which the slice was extracted is modified:

In [ ]:
a[1:3] = [-2,-3]

a

* We can omit any of the three parameters in `M[lower:upper:step]`:

In [ ]:
a[::] # lower, upper, step all take the default values

In [ ]:
a[::2] # step is 2, lower and upper defaults to the beginning and end of the array

In [ ]:
a[:3] # first three elements

In [ ]:
a[3:] # elements from index 3

* Negative indices counts from the end of the array (positive index from the begining):

In [ ]:
a = np.array([1,2,3,4,5])

In [ ]:
a[-1] # the last element in the array

In [ ]:
a[-3:] # the last three elements

* Index slicing works exactly the same way for multidimensional arrays:

In [ ]:
A = np.array([[n+m*10 for n in range(5)] 
              for m in range(5)])
A

In [ ]:
# a block from the original array
A[1:4, 1:4]

In [ ]:
# strides
A[::2, ::2]

## Fancy indexing

Fancy indexing is the name for when an array or list is used in-place of an index: 

In [ ]:
row_indices = [1, 2, 3]
A[row_indices]

In [ ]:
col_indices = [1, 2, -1] # remember, index -1 means the last element
A[row_indices, col_indices]

* We can also index **masks**: 

    - If the index mask is an Numpy array of with data type `bool`, then an element is selected (True) or not (False) depending on the value of the index mask at the position each element: 

In [ ]:
b = np.array([n for n in range(5)])
b

In [ ]:
row_mask = np.array([True, False, True, False, False])
b[row_mask]

* Alternatively:

In [ ]:
# same thing
row_mask = np.array([1,0,1,0,0], dtype=bool)
b[row_mask]

This feature is very useful to conditionally select elements from an array, using for example comparison operators:

In [ ]:
x = np.arange(0, 10, 0.5)
x

In [ ]:
mask = (5 < x)

mask

In [ ]:
x[mask]

Alternatively, we can use the condition (mask) array directly within brackets to index the array

In [ ]:
x[(5 < x)]

---

# Exercises on Indexing

Index slicing is the technical name for the syntax `M[lower:upper:step]` to extract part of an array

## Ex 3.1 

Generate a three-dimensional array of any size containing random numbers taken from an uniform distribution (_guess the numpy function in `np.random`_). Then print out separately the first entry along the three axis (i.e. `x, y, z`)  


#### Hint: Slicing with numpy arrays works quite like Python lists

## Ex 3.2

Create a vector and print out elements in reverse order

#### Hint: Use slicing for this exercise

## Ex 3.3

Generate a $7 \times 7$ matrix and replace all the elements in odd rows and even columns with `1`.

#### Hint: Use slicing to solve this exercise!

#### Note: Take a look at the original matrix, then.

**Use fancy indexing** to get all the elements of the previous matrix that are equals to `1`

## Ex 3.4 

Generate a `10 x 10` matrix of numbers `A`. Then, generate a numpy array of integers in range `1-9`. Pick `5` random values (with no repetition) from this array and use these values to extract rows from the original matrix `A`.

## Ex 3.5 

Repeat the previous exercise but this time extract columns from `A`

## Ex 3.6

Generate an array of numbers from `0` to `20` with step `0.5`. 
Extract all the values greater than a randomly generated number in the same range.

#### Hint: Try to write the condition as an expression and save it to a variable. Then, use this variable in square brackets to index.... this is when the magic happens!